### ISSUE TITLE
updating to a more recent version of ghc

### ISSUE THREAD

**Issue by polymorphicengine** [gelisam/deploy-hint#8]
I have been trying to update your script to work with ghc 8.6.5 here: https://github.com/onthepeakofnormal/deploy-hint/tree/8.6.5
everything compiles fine and the program runs without errors on the with-ghc docker image. I get the following error on the without-ghc docker image:
```
docker cp build-with-ghc:/root/my-program.tar.gz my-program.tar.gz
docker rm -vf run-without-ghc &> /dev/null || true
echo '()' | docker run -i --name run-without-ghc without-ghc /root/my-program/my-program
please type '()':
(\x -> (x,x)) () is:
Left (GhcException "could not execute: gcc")
and now, let's try the Prelude...
id () is:
Left (GhcException "panic! (the 'impossible' happened)\n  (GHC version 8.6.5 for x86_64-unknown-linux):\n\tDynamic linker not initialised\n\nPlease report this as a GHC bug:  http://www.haskell.org/ghc/reportabug\n")
and finally, a library from hackage.
runReader ask () is:
GhcException "panic! (the 'impossible' happened)\n  (GHC version 8.6.5 for x86_64-unknown-linux):\n\tDynamic linker not initialised\n\nPlease report this as a GHC bug:  http://www.haskell.org/ghc/reportabug\n"
```
I'm pretty stuck in trying to understand what I can do to fix it

**Comment by gelisam** [comment]
I haven't seen that one before. Here's one thing it could be.

ghc has a settings file (`/opt/ghc/7.10.3/lib/ghc-7.10.3/settings`) which specifies which external programs to use when constructing an executable: which C compiler to use (e.g. gcc or clang), which linker to use (e.g. ld or gold), that kind of thing. So maybe there is something wrong with that settings file which prevents it from figuring out which linker to use? The fact that it can't find gcc is already an indication that this settings file is incorrect, because it points to a C compiler which is not installed on the target machine.

One thing I don't understand is: why is ghc invoking gcc and the linker here? This is normally used when compiling, but here we are only interpreting. If gcc and a linker are required when interpreting already-compiled code, this would be quite unfortunate since the whole point of this repo is to figure out how to run hint programs on machines which don't have ghc installed. If we manage to do that but still require gcc to be installed, is that still useful?

I know linking is used when TemplateHaskell code is compiled, and thus perhaps also when code containing TemplateHaskell is interpreted. But that doesn't seem to be the case here.

Perhaps some files are missing, and ghc is trying to recreate them? I know that when compiling a library, ghc must generate two sets of files: one which will be used when compiling an executable which uses that library, and one which will be used when loading that library in ghci. Maybe ghc is now storing the second set in a different folder, which we're not copying? Or maybe `cabal build` is now only compiling the first set, and ghc now builds the second set from the first when needed?

In any case, since it works in with-ghc but not in without-ghc, there is a sure-fire way to figure out what's wrong: binary search between the two environments until you find the difference which matters.

**Comment by yaxu** [comment]
I'm not sure when it gets called but there is fake_gcc.sh, which somehow needs to be there but doesn't do anything? and haskell-libs/settings is changed to point to that

**Comment by gelisam** [comment]
Oh right, thanks for reminding me about that! Since ghc calls gcc in order to obtain the name of an `.so` file, I am guessing that a C compiler may choose to mangle the name of the `.so` file it is asked to produce, that gcc doesn't (or doesn't need to in this case), and that ghc asks the C compiler which name it used just before loading a `.so` file with [ldopen](https://man7.org/linux/man-pages/man3/dlopen.3.html). perhaps we now need a fake linker too?

**Comment by yaxu** [comment]
I'm just looking at the my-program.tar.gz that gets produced.. It looks like acme-dont isn't being copied in there. 

Also the settings file has these two lines:
 ("C compiler command", "gcc"),
 ("Haskell CPP command","gcc"),

The regex is looking for /usr/bin/gcc - I guess that needs changing to search for just gcc so that it gets replaced with the path to fake_gcc.sh

**Comment by yaxu** [comment]
Yes it's working now. The lack of acme-dont was expected (@onthepeakofnormal is working on a different library), and fixing the regex from /usr/bin/gcc to just gcc was enough. Hopefully no fake linker needed!

**Comment by polymorphicengine** [comment]
we made some progress here: https://github.com/onthepeakofnormal/deploy-hint/tree/tidal_1.7.5
now the import of tidal works but when interpreting it is looking for some missing c libraries. When I tried adding libdl.so I got a `Segmentation fault (core dumped)` error..

**Comment by yaxu** [comment]
Ok, well I found two ways to fix this..
```
docker run -it  --name build-without-ghc without-ghc /bin/bash
cd my-program
cp /lib/x86_64-linux-gnu/librt.so.1 c-libs/librt.so
cp /lib/x86_64-linux-gnu/libutil.so.1 c-libs/libutil.so
cp /lib/x86_64-linux-gnu/libdl.so.2 c-libs/libdl.so
cp /lib/x86_64-linux-gnu/libpthread.so.0 c-libs/libpthread.so
```
Or just like this:
```
docker run -it  --name build-without-ghc without-ghc /bin/bash
ln -s /lib/x86_64-linux-gnu/librt.so.1 /lib/x86_64-linux-gnu/librt.so
ln -s /lib/x86_64-linux-gnu/libutil.so.1 /lib/x86_64-linux-gnu/libutil.so
ln -s /lib/x86_64-linux-gnu/libdl.so.2 /lib/x86_64-linux-gnu/libdl.so
ln -s /lib/x86_64-linux-gnu/libpthread.so.0 /lib/x86_64-linux-gnu/libpthread.so
```
In both cases the libraries are already there, just not without the numerical extension (which I guess is the library version?).
If I check with `ldd` it points to the right files:
```
ldd my-program 
	linux-vdso.so.1 (0x00007ffd9ab5e000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f495d1a4000)
	libtinfo.so.5 => /root/my-program/./c-libs/libtinfo.so.5 (0x00007f495cf7a000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f495cd72000)
	libutil.so.1 => /lib/x86_64-linux-gnu/libutil.so.1 (0x00007f495cb6f000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f495c96b000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f495c74e000)
	libgmp.so.10 => /root/my-program/./c-libs/libgmp.so.10 (0x00007f495c4cb000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f495c120000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f495d4a5000)
```
.. so I guess this is something to do with the interpreter not pointing to particular library versions.

Well this is well out of my expertise.. But maybe these are all really standard system libraries that any 'real' distro will have.. It runs fine on my system anyway! So I don't know if we actually want to copy these files in or not.

**Comment by yaxu** [comment]
Looking further the segfaults only happen when we copy the c libs from with-ghc to without-ghc. 

with-ghc has more recent libraries than without-ghc. without-ghc is based on debian jessie, my guess is that with-ghc is based on stretch.

So it seems we should use the system version of these libraries and not copy them into c-libs?

If that's true, then the only issue is that without-ghc doesn't have symlinks for librt.so to point at librt.so.1, etc. I don't know why they aren't already there. Otherwise it works!

**Comment by gelisam** [comment]
I think this is where @juhp's expertise comes in :)

**Comment by juhp** [comment]
Hmm the unversioned symlinks live in the libc6-dev package, but that might well pull in gcc...?
(Normally on Linux we always dynlink to versioned .so libs, so they are only needed at runtime, and the unversioned symlinks are only used in development.)

I think your symlinks hack could work, @yaxu, at least for now.

(It might be better to use the same base container image for both for consistency, though sounds like you are close already.)

**Comment by yaxu** [comment]
Thanks both! We can try switching to stretch for the target, although it's also nice to see that this works across versions. 

So is it a problem that hint is using unversioned libs?

Would it be better to make a static binary? 

I've copied @onthepeakofnormal's work here and made it relocatable, it works! https://github.com/tidalcycles/tidal-deploy/tree/main/from-deploy-hint

We can make a quick pr here as well later, and can include removing the customized hint which isn't used any more afaict.

I guess the same approach will work on macos, Windows will be a different matter I suppose..
